### PR TITLE
start in on validation policies

### PR DESCRIPTION
**Pull request by zakird**
WIP. Gonna add some tests, but I think the high-level idea is there.

### CODE REVIEW

File: zschema/keys.py [hunk 1]
```diff
@@ -168,6 +171,43 @@ def key_to_string(o):
         else:
             return o.to_string()
 
+    @staticmethod
+    def _handle_validation(policy, e):
```
Comment (zakird) [inline zschema/keys.py]: This should probably get a better name in retrospect.

**Comment by zakird** [comment]
This has a bug .validate on leaves need to be changed or they'll use the parent subclasses record not their own.

File: zschema/compounds.py [hunk 2]
```diff
@@ -215,16 +226,26 @@ def to_dict(self):
...
-                raise DataValidationException("%s: %s is not a valid subkey" %
-                                              (name, subkey))
-            else:
-                self.definition[subkey].validate(subkey, subvalue)
+            try:
+                if not self.allow_unknown and subkey not in self.definition:
+                    raise DataValidationException("%s: %s is not a valid subkey" %
+                                                  (name, subkey))
+                else:
+                    self.definition[subkey].validate(subkey, subvalue,
```
Comment (justinbastress) [inline zschema/compounds.py]: This raises a `KeyError` if both `self.allow_unknown` and `subkey not in self.definition` (not a new problem); it seems the intent is `elif subkey in self.defnition`?

File: zschema/keys.py (hunk 1 above)
Comment (justinbastress) [reply zschema/keys.py]: If it was go, you could define a `policy.validation_failure(e)`; but as it stands, `_handle_validation_failure(policy, e)` would make its intent clearer IMO.

File: zschema/keys.py [hunk 3]
```diff
@@ -168,6 +171,43 @@ def key_to_string(o):
...
+            raise Exception("Invalid policy for %s: %s" % (name, policy))
+
+    def _calculate_policy(self, name, policy, parent_policy):
+        # Validation policies can be enforced several ways in the following
+        # priority order:
+        #
+        # 1) explicit policy passed into .validate
+        # 2) policy defined in the object:
+        #       a) policy at object creation
+        #       b) default or that field type
```
Comment (justinbastress) [inline zschema/keys.py]: or -> on?

File: zschema/keys.py (hunk 1 above)
Comment (zakird) [reply zschema/keys.py]: ✔️

File: zschema/keys.py (hunk 3 above)
Comment (zakird) [reply zschema/keys.py]: ✔️

File: zschema/compounds.py (hunk 2 above)
Comment (zakird) [reply zschema/compounds.py]: I don't think that's right. The code is:
` if not self.allow_unknown and subkey not in self.definition:`

The key part being if **not** allow_unknown.

Comment (zakird) [reply zschema/compounds.py]: ok, I see what you mean (although I think not what you wrote 😋). fixed.

**Comment by zakird** [comment]
alright @justinbastress I think I've fixed up all the bugs. Should be good to review. I'm going to write some examples/tests and make sure this meets @cdzombak's biggest dreams and desires.

**Comment by zakird** [comment]
Alright! Added some documentation here: https://github.com/zmap/zschema/wiki/Validating-Data

File: zschema/compounds.py [hunk 4]
```diff
@@ -58,12 +58,21 @@ def docs_es(self, parent_category=None):
...
         return retv
 
-    def validate(self, name, value):
+    def validate(self, name, value, policy=_NO_ARG, parent_policy=_NO_ARG):
+        calculated_policy = self._calculate_policy(name, policy, parent_policy)
         if type(value) != list:
-            raise DataValidationException("%s: %s is not a list",
-                                          name, str(value))
+            try:
+                raise DataValidationException("%s: %s is not a list",
```
Comment (justinbastress) [inline zschema/compounds.py]: Should be `DataValidationException(fmt % (...))`, not `DataValidationException(fmt, ...)`, right?

**Comment by zakird** [comment]
@justinbastress fixed

File: zschema/keys.py [hunk 5]
```diff
@@ -168,6 +171,43 @@ def key_to_string(o):
...
+            raise e
+        elif policy == "warn":
+            logging.warn(e.message)
+        elif policy == "ignore":
+            pass
+        else:
+            raise Exception("Invalid validation policy. Must be one of: error, warn, ignore")
+
+    @staticmethod
+    def _validate_policy(name, policy):
```
Comment (justinbastress) [inline zschema/keys.py]: Reading through the code, I'd initially assumed `validate_policy` validates that the policy was being followed.

What about `validation_policy_enum(policy, name=None)`, by analogy to e.g. `int()`? (or `validation_policy_id` or just `validation_policy` etc)

Then you could also use `type=validation_policy_enum` in the `argparse`.

File: zschema/compounds.py [hunk 6]
```diff
@@ -215,16 +226,26 @@ def to_dict(self):
...
 
-    def validate(self, name, value):
-        if type(value) != dict:
-            raise DataValidationException("%s: %s is not a dict",
-                                          name, str(value))
+    def validate(self, name, value, policy=_NO_ARG, parent_policy=_NO_ARG):
+        calculated_policy = self._calculate_policy(name, policy, parent_policy)
+        try:
+            if not isinstance(value, dict):
+                raise DataValidationException("%s: %s is not a dict",
```
Comment (justinbastress) [inline zschema/compounds.py]: Format string again

**Comment by zakird** [comment]
I'm going to fix the formatting string and merge so that @cdzombak can use. I'll fix up enum method when I add the CLI endpoint here in a moment.

**Pull request merged**